# Beginner's guide to scRNA-seq - GI edition

## Day 1: Tue May 27th 2025

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [Introduction to R/Rstudio/Seurat](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-1/1%20-%20Introduction.pdf) [[Recording](https://drive.google.com/drive/folders/1h5XdDTF2iCdJnKA_IW3qVucDCBz0jIwu)]
        -   How to navigate Rstudio
        -   Fundamentals and data wrangling in R
        -   Working with a Seurat object
    -   [Basic sc-RNAseq workflow](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-1/2-The%20Basic%20scRNAseq%20Analysis%20Workflow.pdf) [[Recording]()]
        -   Main steps for scRNAseq workflow
        -   Intuition behind each step
    -   [Data Download](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-1/3-download-data.html) [[Recording]()]
        - Download the data needed for this workshop
        - Go over the basics of a Seurat object
    -   [Normalization & HVG selection](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-1/4-norm-hvg.html) [[Recording]()]
        -   Why and how to normalize sc-RNAseq data
        -   Why and how to select highly variable genes
    -   [Quality Control](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-1/5-QC.html) [[Recording]()]
        -   How to carry out QC analysis
        -   What parameters do we need to look at and how to interpret them
        -   Computing doublet scores
        -   Tips and tricks during the QC process

## Day 2: Tue June 3rd 2025

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [PCA, Integration & KNN graphs](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-2/6-PCA_Harmony_kNN.html) [[Recording]()]
        -   Brief overview on PCA and how to use it for sc-RNAseq
        -   KNN-graph representation of the data
        -   Correcting batch effects with Harmony
    -   [Clustering](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-2/7-Clustering.html) [[recording]()]
        -   How to cluster sc-RNAseq data
        -   What algorithms can we use
        -   How to assess if a clustering resolution is good
    -   [Differential Gene Expression & Level 1 Annotation](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-2/8-dge-annotlvl1.html) [[recording]()]
        -   How to compute differentially expressed genes between clusters
        -   How to evaluate differential expression statistics
        -   Reference-based cell type annotation
        -   Manual cell type annotation
    -   [Subclustering & Level 2 Annotation](http://htmlpreview.github.io/?https://github.com/CellDiscoveryNetwork/beginner-workshop-gi-5-25/blob/main/day-2/9-compositional_analysis.html) [[recording]()]
        -   What does level 2 annotation mean and why we need it
        -   How to iteratively annotate a dataset

### Exit Forms
TBD
