# Beginner's guide to scRNA-seq - GI edition

## Day 1: Tue May 27th 2025

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [Introduction to R/Rstudio/Seurat]() [[Recording]()]
        -   How to navigate Rstudio
        -   Fundamentals and data wrangling in R
        -   Working with a Seurat object
    -   [Basic sc-RNAseq workflow]() [[Recording]()]
        -   Main steps for scRNAseq workflow
        -   Intuition behind each step
    -   [Data Download]()[[Recording]()]
        - Download the data needed for this workshop
        - Go over the basics of a Seurat object
    -   [Normalization & HVG selection]()[[Recording]()]
        -   Why and how to normalize sc-RNAseq data
        -   Why and how to select highly variable genes
    -   [Quality Control]() [[Recording]()]
        -   How to carry out QC analysis
        -   What parameters do we need to look at and how to interpret them
        -   Computing doublet scores
        -   Tips and tricks during the QC process

## Day 2: Tue June 3rd 2025

-   **Where**: Karp 6 Conference Room
-   **Course Material**:
    -   [PCA, Integration & KNN graphs]()[[Recording]()]
        -   Brief overview on PCA and how to use it for sc-RNAseq
        -   KNN-graph representation of the data
        -   Correcting batch effects with Harmony
    -   [Clustering]() [[recording]()]
        -   How to cluster sc-RNAseq data
        -   What algorithms can we use
        -   How to assess if a clustering resolution is good
    -   [Differential Gene Expression & Level 1 Annotation]() [[recording]()]
        -   How to compute differentially expressed genes between clusters
        -   How to evaluate differential expression statistics
        -   Reference-based cell type annotation
        -   Manual cell type annotation
    -   [Subclustering & Level 2 Annotation]() [[recording]()]
        -   What does level 2 annotation mean and why we need it
        -   How to iteratively annotate a dataset

### Exit Forms
TBD
